README.md
# cordova-plugin-mauron85-background-geolocation

## Description

Cross-platform geolocation for Cordova / PhoneGap with battery-saving "circular region monitoring" and "stop detection".

Plugin is both foreground and background geolocation service. It is far more battery and data efficient then html5 geolocation or cordova-geolocation plugin. But it can be used together with other geolocation providers (eg. html5 navigator.geolocation).

On Android you can choose from two location location providers:
* **ANDROID_DISTANCE_FILTER_PROVIDER** (forked from [cordova-plugin-background-geolocation](https://github.com/christocracy/cordova-plugin-background-geolocation))
* **ANDROID_ACTIVITY_PROVIDER**

See wiki [Which provider should I use?](https://github.com/mauron85/cordova-plugin-background-geolocation/blob/master/PROVIDERS.md) for more information about providers.

## Submitting issues

All new issues should follow instructions in [ISSUE_TEMPLATE.md](https://raw.githubusercontent.com/mauron85/cordova-plugin-background-geolocation/master/ISSUE_TEMPLATE.md). Properly filled issue report will significantly reduce number of follow up questions and decrease issue resolve time.

## Semantic Versioning
This plugin is following semantic versioning as defined http://semver.org

## Migration to 2.0

Warning: `option.url` for posting locations is very experimental and missing features like remote
server synchronization. Location database can get very big as currently there is no cleaning mechanism.
Use it at own risk. Proper server synchronization will be implemented in version 3.0.

As version 2.0 platform support for Windows Phone 8 was removed.
Some incompatible changes were introduced:

* option `stopOnTerminate` defaults to true
* option `locationService` renamed to `locationProvider`
* android providers are now **ANDROID_DISTANCE_FILTER_PROVIDER** and **ANDROID_ACTIVITY_PROVIDER**
* removed `locationTimeout` option (use `interval` in milliseconds instead)
* `notificationIcon` was replaced with two separate options (`notificationIconSmall` and `notificationIconLarge`)
* js object backgroundGeoLocation is deprecated use `backgroundGeolocation` instead
* iOS foreground mode witch automatic background mode switch
* iOS [switchMode][] allows to switch between foreground and background mode
* setPace on iOS is deprecated use switchMode instead


## Installing the plugin

```
cordova plugin add cordova-plugin-mauron85-background-geolocation
```

## Registering plugin for Adobe® PhoneGap™ Build

This plugin should work with Adobe® PhoneGap™ Build without any modification.
To register plugin add following line into your config.xml:

```
<plugin name="cordova-plugin-mauron85-background-geolocation" spec="^2.2.0"/>
```

NOTE: If you're using *hydration*, you have to download and reinstall your app with every new version of the plugin, as plugins are not updated.

## Compilation

### Android
You will need to ensure that you have installed the following items through the Android SDK Manager:

Name                       | Version
-------------------------- | -------
Android SDK Tools          | 24.4.1
Android SDK Platform-tools | 23.1
Android SDK Build-tools    | 23.0.1
Android Support Repository | 25
Android Support Library    | 23.1.1
Google Play Services       | 29
Google Repository          | 24

## Quick Example

```javascript
document.addEventListener('deviceready', onDeviceReady, false);

function onDeviceReady () {

    /**
    * This callback will be executed every time a geolocation is recorded in the background.
    */
    var callbackFn = function(location) {
        console.log('[js] BackgroundGeolocation callback:  ' + location.latitude + ',' + location.longitude);

        // Do your HTTP request here to POST location to your server.
        // jQuery.post(url, JSON.stringify(location));

        /*
        IMPORTANT:  You must execute the finish method here to inform the native plugin that you're finished,
        and the background-task may be completed.  You must do this regardless if your HTTP request is successful or not.
        IF YOU DON'T, ios will CRASH YOUR APP for spending too much time in the background.
        */
        backgroundGeolocation.finish();
    };

    var failureFn = function(error) {
        console.log('BackgroundGeolocation error');
    };

    // BackgroundGeolocation is highly configurable. See platform specific configuration options
    backgroundGeolocation.configure(callbackFn, failureFn, {
        desiredAccuracy: 10,
        stationaryRadius: 20,
        distanceFilter: 30,
        interval: 60000
    });

    // Turn ON the background-geolocation system.  The user will be tracked whenever they suspend the app.
    backgroundGeolocation.start();

    // If you wish to turn OFF background-tracking, call the #stop method.
    // backgroundGeolocation.stop();
}
```

## Example Application

Checkout repository [cordova-plugin-background-geolocation-example](https://github.com/mauron85/cordova-plugin-background-geolocation-example).

## API

### backgroundGeolocation.configure(success, fail, option)

Parameter | Type | Platform     | Description
--------- | ---- | ------------ | -----------
`success` | `Function` | all | Callback to be executed every time a geolocation is recorded in the background.
`fail` | `Function` | all | Callback to be executed every time a geolocation error occurs.
`option` | `JSON Object` | all |
`option.desiredAccuracy` | `Number` | all | Desired accuracy in meters. Possible values [0, 10, 100, 1000]. The lower the number, the more power devoted to GeoLocation resulting in higher accuracy readings.  1000 results in lowest power drain and least accurate readings. **@see** [Apple docs](https://developer.apple.com/library/ios/documentation/CoreLocation/Reference/CLLocationManager_Class/CLLocationManager/CLLocationManager.html#//apple_ref/occ/instp/CLLocationManager/desiredAccuracy)
`option.stationaryRadius` | `Number` | all | Stationary radius in meters. When stopped, the minimum distance the device must move beyond the stationary location for aggressive background-tracking to engage.
`option.debug` | `Boolean` | all | When enabled, the plugin will emit sounds for life-cycle events of background-geolocation! See debugging sounds table.
`option.distanceFilter` | `Number` | all | The minimum distance (measured in meters) a device must move horizontally before an update event is generated. **@see** [Apple docs](https://developer.apple.com/library/ios/documentation/CoreLocation/Reference/CLLocationManager_Class/CLLocationManager/CLLocationManager.html#//apple_ref/occ/instp/CLLocationManager/distanceFilter).
`option.stopOnTerminate` | `Boolean` | iOS, Android | Enable this in order to force a stop() when the application terminated (e.g. on iOS, double-tap home button, swipe away the app). (default true)
`option.startOnBoot` | `Boolean` | Android | Start background service on device boot. (default false)
`option.startForeground` | `Boolean` | Android | If false location service will not be started in foreground and no notification will be shown. (default true)
`option.interval` | `Number` | Android | The minimum time interval between location updates in milliseconds. **@see** [Android docs](http://developer.android.com/reference/android/location/LocationManager.html#requestLocationUpdates(long,%20float,%20android.location.Criteria,%20android.app.PendingIntent) for more information.
`option.notificationTitle` | `String` optional | Android | Custom notification title in the drawer.
`option.notificationText` | `String` optional | Android | Custom notification text in the drawer.
`option.notificationIconColor` | `String` optional| Android | The accent color to use for notification. Eg. **#4CAF50**.
`option.notificationIconLarge` | `String` optional | Android | The filename of a custom notification icon. See android quirks.
`option.notificationIconSmall` | `String` optional | Android | The filename of a custom notification icon. See android quirks.
`option.locationProvider` | `Number` | Android | Set location provider **@see** [wiki](https://github.com/mauron85/cordova-plugin-background-geolocation/wiki/Android-providers)
`option.activityType` | `String` | iOS | [AutomotiveNavigation, OtherNavigation, Fitness, Other] Presumably, this affects iOS GPS algorithm. **@see** [Apple docs](https://developer.apple.com/library/ios/documentation/CoreLocation/Reference/CLLocationManager_Class/CLLocationManager/CLLocationManager.html#//apple_ref/occ/instp/CLLocationManager/activityType) for more information
`option.url` | `String` | iOS, Android | Server url where to send HTTP POST with recorded locations
`option.syncUrl` | `String` | iOS | Server url where to send fail to post locations (if left blank, same option.url will be used)
`option.syncThreshold` | `Number` | iOS | Specifies how many previously failed locations will be sent to server at once (default: 100)
`option.httpHeaders` | `Object` | iOS, Android | Optional HTTP headers sent along in HTTP request
`option.saveBatteryOnBackground` | `Boolean` | iOS | If enabled it will automatically switch to less accurate significant changes and region monitory when in background (default)
`options.maxLocations` | `Number` | iOS | Limit maximum number of locations stored into db (default: 10000)

Following options are specific to provider as defined by locationProvider option
### ANDROID_ACTIVITY_PROVIDER provider options

Parameter | Type | Platform     | Description
--------- | ---- | ------------ | -----------
`option.interval` | `Number` | Android | Rate in milliseconds at which your app prefers to receive location updates. @see [android docs](https://developers.google.com/android/reference/com/google/android/gms/location/LocationRequest.html#getInterval())
`option.fastestInterval` | `Number` | Android | Fastest rate in milliseconds at which your app can handle location updates. **@see** [android  docs](https://developers.google.com/android/reference/com/google/android/gms/location/LocationRequest.html#getFastestInterval()).
`option.activitiesInterval` | `Number` | Android | Rate in milliseconds at which activity recognition occurs. Larger values will result in fewer activity detections while improving battery life.
`option.stopOnStillActivity` | `Boolean` | Android | stop() is forced, when the STILL activity is detected (default is true)

Success callback will be called with one argument - location object, which tries to mimic w3c [Coordinates interface](http://dev.w3.org/geo/api/spec-source.html#coordinates_interface).

Callback parameter | Type | Description
------------------ | ---- | -----------
`locationId` | `Number` | ID of location as stored in DB (or null)
`provider` | `String` | gps, network, passive or fused
`locationProvider` | `Number` | Location provider
`debug` | `Boolean` | true if location recorded as part of debug
`time` | `Number` |UTC time of this fix, in milliseconds since January 1, 1970.
`latitude` | `Number` | latitude, in degrees.
`longitude` | `Number` | longitude, in degrees.
`accuracy` | `Number` | estimated accuracy of this location, in meters.
`speed` | `Number` | speed if it is available, in meters/second over ground.
`altitude` | `Number` | altitude if available, in meters above the WGS 84 reference ellipsoid.
`bearing` | `Number` | bearing, in degrees.


### backgroundGeolocation.start()
Platform: iOS, Android

Start background geolocation.

### backgroundGeolocation.stop()
Platform: iOS, Android

Stop background geolocation.

### backgroundGeolocation.isLocationEnabled(success, fail)
Platform: iOS, Android

One time check for status of location services. In case of error, fail callback will be executed.

Success callback parameter | Type | Description
-------------------------- | ---- | -----------
`enabled` | `Boolean` | true/false (true when location services are enabled)

### backgroundGeolocation.showAppSettings()
Platform: Android >= 6, iOS >= 8.0

Show app settings to allow change of app location permissions.

### backgroundGeolocation.showLocationSettings()
Platform: iOS, Android

Show system settings to allow configuration of current location sources.

### backgroundGeolocation.watchLocationMode(success, fail)
Platform: iOS, Android

Method can be used to detect user changes in location services settings.
If user enable or disable location services then success callback will be executed.
In case or error (SettingNotFoundException) fail callback will be executed.

Success callback parameter | Type | Description
-------------------------- | ---- | -----------
`enabled` | `Boolean` | true/false (true when location services are enabled)

### backgroundGeolocation.stopWatchingLocationMode()
Platform: iOS, Android

Stop watching for location mode changes.

### backgroundGeolocation.getLocations(success, fail)
Platform: iOS, Android

Method will return all stored locations.

CHANGES.md
## Changelog

### [2.2.0] - unreleased
### Added
- Android showAppSettings
- database logging (getLogEntries)
- autosync locations to server with configurable threshold
- maxLocations limit
- method getValidLocations

### Changed
- Locations stored into db at all times

### Fixed
- Android permission error code: 2
- Android on start err callback instead configure err callback

### [2.1.2] - 2016-06-23
### Fixed
- iOS database not created

### [2.1.1] - private release
### Fixed
- iOS switching mode

### [2.1.0] - private release
### Added
- iOS option saveBatteryOnBackground
- iOS time validation rule for location

### [2.0.0] - 2016-06-17
### Fixed
- iOS prevent unintentional start when in background
- Android Destroy Existing Provider Before Creating New One (#94)

### [2.0.0-rc.3] - 2016-06-13
#### Fixed
- iOS memory leak

### [2.0.0-rc.1] - 2016-06-13
#### Changed
- Android notificationIcon option split into small and large!!!
- Android stopOnTerminate defaults to true
- Android option locationService renamed to locationProvider
- Android providers renamed (see README.md)
- Android bugfixing
- SampleApp moved into separate repo
- deprecated backgroundGeoLocation
- iOS split cordova specific code to allow code sharing with react-native-background-geolocation
- desiredAccuracy map any number
- Android locationTimeout option renamed to interval
- iOS switchMode (formerly setPace)

#### Added
- Android startOnBoot option
- Android startForeground option
- iOS, Android http posting of locations (options url and httpHeaders)
- iOS showLocationSettings
- iOS showAppSettings
- iOS isLocationEnabled
- iOS getLocations
- iOS deleteLocation
- iOS deleteAllLocations
- iOS foreground mode

#### Removed
- WP8 platform
- Android deprecated window.plugins.backgroundGeoLocation

### [1.0.2] - 2016-06-09
#### Fixed
- iOS queued locations are send FIFO (before fix LIFO)

### [1.0.1] - 2016-06-03
#### Fixed
- iOS7 crash on start
- iOS attempt to fix #46 and #39

### [1.0.0] - 2016-06-01
#### Added
- Android ANDROID_FUSED_LOCATION stopOnStillActivity (enhancement #69)

### [0.9.6] - 2016-04-07
#### Fixed
- Android ANDROID_FUSED_LOCATION fixing crash on start
- Android ANDROID_FUSED_LOCATION unregisterReceiver on destroy

### [0.9.5] - 2016-04-05
#### Fixed
- Android ANDROID_FUSED_LOCATION startTracking when STILL after app has started

### [0.9.4] - 2016-01-31
#### Fixed
- Android 6.0 permissions issue #21

### [0.9.3] - 2016-01-29
#### Fixed
- iOS cordova 6 compilation error
- iOS fix for iOS 9

#### Changes
- iOS removing cordova-plugin-geolocation dependency
- iOS user prompt for using location services
- iOS error callback when location services are disabled
- iOS error callback when user denied location tracking
- iOS adding error callbacks to SampleApp

### [0.9.2] - 2016-01-29
#### Fixed
- iOS temporarily using cordova-plugin-geolocation-ios9-fix to fix issues with iOS9
- iOS fixing SampleApp indexedDB issues

### [0.9.1] - 2015-12-18
#### Fixed
- Android ANDROID_FUSED_LOCATION fix config setActivitiesInterval

### [0.9.0] - 2015-12-18
#### Changed
- Android ANDROID_FUSED_LOCATION using ActivityRecognition (saving battery)

### [0.8.3] - 2015-12-18
#### Fixed
- Android fixing crash on exit

### [0.8.2] - 2015-12-18
#### Fixed
- Android fixing #9 - immediate bg service crash

### [0.8.1] - 2015-12-15
#### Fixed
- Android fixing #9

### [0.8.0] - 2015-12-15 (Merry XMas Edition :-)
#### Fixed
- Android persist location when main activity was killed

#### Changed
- Android persisting position when debug is on

### [0.7.3] - 2015-11-06
#### Fixed
- Android issue #11

### [0.7.2] - 2015-10-21
#### Fixed
- iOS fixing plugin dependencies (build)
- iOS related fixes for SampleApp

### [0.7.1] - 2015-10-21
#### Changed
- Android ANDROID_FUSED_LOCATION ditching setSmallestDisplacement(stationaryRadius) (seems buggy)

### [0.7.0] - 2015-10-21
#### Changed
- Android deprecating config option.interval
- Android allow run in background for FusedLocationService (wakeLock)
- Android will try to persist locations when main activity is killed
- Android new methods: (getLocations, deleteLocation, deleteAllLocations)
- Android stop exporting implicit intents (security)
- SampleApp updates

### [0.6.0] - 2015-10-17
#### Changed
- deprecating window.plugins clobber
- SampleApp updates

#### Added
- Android showLocationSettings and watchLocationMode

### [0.5.4] - 2015-10-13
#### Changed
- Android only cosmetic changes, but we need stable base

### [0.5.3] - 2015-10-12
#### Changed
- Android not setting any default notificationIcon and notificationIconColor.
- Android refactoring
- Android updated SampleApp

### [0.5.2] - 2015-10-12
#### Fixed
- Android fixing FusedLocationService start and crash on stop

### [0.5.1] - 2015-10-12
#### Fixed
- Android fix return types
- Android fix #3 NotificationBuilder.setColor method not present in API Level <21

#### Changed
- Android replacing Notication.Builder for NotificationCompat.Builder
- SampleApp can send position to server.
- SampleApp offline mode (IndexedDB)

#### Removed
- Android unnecessary plugins
- Docs: removing instructions to enable cordova geolocation in foreground
 and user accept location services

### [0.5.0] - 2015-10-10
#### Changed
- Android FusedLocationService
- Android package names reverted
- Android configuration refactored
- WP8 merged improvements

#### Removed
- Android unused classes
- All removing deprecated url, params, headers

### [0.4.3] - 2015-10-09
#### Added
- Android Add icon color parameter

#### Changed
- Changed the plugin.xml dependencies to the new NPM-based plugin syntax
- updated SampleApp

### [0.4.2] - 2015-09-30
#### Added
- Android open activity when notification clicked [69989e79a8a67485fc88463eec8d69bb713c2dbe](https://github.com/erikkemperman/cordova-plugin-background-geolocation/commit/69989e79a8a67485fc88463eec8d69bb713c2dbe)

#### Fixed
- Android duplicate desiredAccuracy extra

src/android/Config.java
/*
According to apache license

This is fork of christocracy cordova-plugin-background-geolocation plugin
https://github.com/christocracy/cordova-plugin-background-geolocation

This is a new class
*/

package com.marianhello.cordova.bgloc;

import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONException;
import android.os.Parcel;
import android.os.Parcelable;

import java.util.HashMap;
import java.util.Map;

/**
 * Config class
 */
public class Config implements Parcelable
{
    private float stationaryRadius = 50;
    private Integer distanceFilter = 500;
    private Integer locationTimeout = 60;
    private Integer desiredAccuracy = 100;
    private Boolean debugging = false;
    private String notificationTitle = "Background tracking";
    private String notificationText = "ENABLED";
    private String activityType; //not used
    private Boolean stopOnTerminate = false;
    private String notificationIcon;
    private String notificationIconColor;
    private ServiceProvider serviceProvider = ServiceProvider.ANDROID_DISTANCE_FILTER;
    private Integer interval = 600000; //milliseconds
    private Integer fastestInterval = 120000; //milliseconds
    private Integer activitiesInterval = 1000; //milliseconds
    private String url;
    private String headers;
    private String params;
    private String oauthUrl;
    private String accessToken;
    private String refreshToken;
    private String clientId;
    private String username;
    private String password;

    public int describeContents() {
        return 0;
    }

    // write your object's data to the passed-in Parcel
    public void writeToParcel(Parcel out, int flags) {
        out.writeFloat(getStationaryRadius());
        out.writeInt(getDistanceFilter());
        out.writeInt(getLocationTimeout());
        out.writeInt(getDesiredAccuracy());
        out.writeValue(isDebugging());
        out.writeString(getNotificationTitle());
        out.writeString(getNotificationText());
        out.writeString(getActivityType());
        out.writeValue(getStopOnTerminate());
        out.writeString(getNotificationIcon());
        out.writeString(getNotificationIconColor());
        out.writeInt(getServiceProvider().asInt());
        out.writeInt(getInterval());
        out.writeInt(getFastestInterval());
        out.writeInt(getActivitiesInterval());
        out.writeString(getUrl());
        out.writeString(getHeaders());
        out.writeString(getParams());
        out.writeString(getOauthUrl());
        out.writeString(getAccessToken());
        out.writeString(getRefreshToken());
        out.writeString(getClientId());
        out.writeString(getUsername());
        out.writeString(getPassword());
    }

    public static final Parcelable.Creator<Config> CREATOR
            = new Parcelable.Creator<Config>() {
        public Config createFromParcel(Parcel in) {
            return new Config(in);
        }

        public Config[] newArray(int size) {
            return new Config[size];
        }
    };

    public Config () {

    }

    public Config(String json) throws JSONException {
        JSONObject jsonObject = new JSONObject(json);

        if (jsonObject.has("stationaryRadius") && !jsonObject.isNull("stationaryRadius"))
            setStationaryRadius((float) jsonObject.getDouble("stationaryRadius"));

        if (jsonObject.has("distanceFilter") && !jsonObject.isNull("distanceFilter"))
            setDistanceFilter(jsonObject.getInt("distanceFilter"));

        if (jsonObject.has("locationTimeout") && !jsonObject.isNull("locationTimeout"))
            setLocationTimeout(jsonObject.getInt("locationTimeout"));

        if (jsonObject.has("desiredAccuracy") && !jsonObject.isNull("desiredAccuracy"))
            setDesiredAccuracy(jsonObject.getInt("desiredAccuracy"));

        if (jsonObject.has("debugging") && !jsonObject.isNull("debugging"))
            setDebugging(jsonObject.getBoolean("debugging"));

        if (jsonObject.has("notificationTitle") && !jsonObject.isNull("notificationTitle"))
            setNotificationTitle(jsonObject.getString("notificationTitle"));

        if (jsonObject.has("notificationText") && !jsonObject.isNull("notificationText"))
            setNotificationText(jsonObject.getString("notificationText"));

        if (jsonObject.has("activityType") && !jsonObject.isNull("activityType"))
            setActivityType(jsonObject.getString("activityType"));

        if (jsonObject.has("stopOnTerminate") && !jsonObject.isNull("stopOnTerminate"))
            setStopOnTerminate(jsonObject.getBoolean("stopOnTerminate"));

        if (jsonObject.has("notificationIcon") && !jsonObject.isNull("notificationIcon"))
            setNotificationIcon(jsonObject.getString("notificationIcon"));

        if (jsonObject.has("notificationIconColor") && !jsonObject.isNull("notificationIconColor"))
            setNotificationIconColor(jsonObject.getString("notificationIconColor"));

        if (jsonObject.has("serviceProvider") && !jsonObject.isNull("serviceProvider")) {
            String serviceProvider = jsonObject.getString("serviceProvider");
            setServiceProvider(ServiceProvider.valueOf(serviceProvider));
        }

        if (jsonObject.has("interval") && !jsonObject.isNull("interval"))
            setInterval(jsonObject.getInt("interval"));

        if (jsonObject.has("fastestInterval") && !jsonObject.isNull("fastestInterval"))
            setFastestInterval(jsonObject.getInt("fastestInterval"));

        if (jsonObject.has("activitiesInterval") && !jsonObject.isNull("activitiesInterval"))
            setActivitiesInterval(jsonObject.getInt("activitiesInterval"));

        if (jsonObject.has("url") && !jsonObject.isNull("url"))
            setUrl(jsonObject.getString("url"));

        if (jsonObject.has("headers") && !jsonObject.isNull("headers"))
            setHeaders(jsonObject.getString("headers"));

        if (jsonObject.has("params") && !jsonObject.isNull("params"))
            setParams(jsonObject.getString("params"));

        if (jsonObject.has("oauthUrl") && !jsonObject.isNull("oauthUrl"))
            setOauthUrl(jsonObject.getString("oauthUrl"));
        if (jsonObject.has("accessToken") && !jsonObject.isNull("accessToken"))
            setAccessToken(jsonObject.getString("accessToken"));
        if (jsonObject.has("refreshToken") && !jsonObject.isNull("refreshToken"))
            setRefreshToken(jsonObject.getString("refreshToken"));
        if (jsonObject.has("clientId") && !jsonObject.isNull("clientId"))
            setClientId(jsonObject.getString("clientId"));
        if (jsonObject.has("username") && !jsonObject.isNull("username"))
            setUsername(jsonObject.getString("username"));
        if (jsonObject.has("password") && !jsonObject.isNull("password"))
            setPassword(jsonObject.getString("password"));

    }

    private Config(Parcel in) {
        setStationaryRadius(in.readFloat());
        setDistanceFilter(in.readInt());
        setLocationTimeout(in.readInt());
        setDesiredAccuracy(in.readInt());
        setDebugging((Boolean) in.readValue(null));
        setNotificationTitle(in.readString());
        setNotificationText(in.readString());
        setActivityType(in.readString());
        setStopOnTerminate((Boolean) in.readValue(null));
        setNotificationIcon(in.readString());
        setNotificationIconColor(in.readString());
        setServiceProvider(in.readInt());
        setInterval(in.readInt());
        setFastestInterval(in.readInt());
        setActivitiesInterval(in.readInt());
        setUrl(in.readString());
        setHeaders(in.readString());
        setParams(in.readString());
        setOauthUrl(in.readString());
        setAccessToken(in.readString());
        setRefreshToken(in.readString());
        setClientId(in.readString());
        setUsername(in.readString());
        setPassword(in.readString());
    }

    public float getStationaryRadius() {
        return stationaryRadius;
    }

    public void setStationaryRadius(float stationaryRadius) {
        this.stationaryRadius = stationaryRadius;
    }

    public Integer getDesiredAccuracy() {
        return desiredAccuracy;
    }

    public void setDesiredAccuracy(Integer desiredAccuracy) {
        this.desiredAccuracy = desiredAccuracy;
    }

    public Integer getDistanceFilter() {
        return distanceFilter;
    }

    public void setDistanceFilter(Integer distanceFilter) {
        this.distanceFilter = distanceFilter;
    }

    public Integer getLocationTimeout() {
        return locationTimeout;
    }

    public void setLocationTimeout(Integer locationTimeout) {
        this.locationTimeout = locationTimeout;
    }

    public Boolean isDebugging() {
        return debugging;
    }

    public void setDebugging(Boolean debugging) {
        this.debugging = debugging;
    }

    public String getNotificationIconColor() {
        return notificationIconColor;
    }

    public void setNotificationIconColor(String notificationIconColor) {
        if (!"null".equals(notificationIconColor)) {
            this.notificationIconColor = notificationIconColor;
        }
    }

    public String getNotificationIcon() {
        return notificationIcon;
    }

    public void setNotificationIcon(String notificationIcon) {
        if (!"null".equals(notificationIcon)) {
            this.notificationIcon = notificationIcon;
        }
    }

    public String getNotificationTitle() {
        return notificationTitle;
    }

    public void setNotificationTitle(String notificationTitle) {
        this.notificationTitle = notificationTitle;
    }

    public String getNotificationText() {
        return notificationText;
    }

    public void setNotificationText(String notificationText) {
        this.notificationText = notificationText;
    }

    public Boolean getStopOnTerminate() {
        return stopOnTerminate;
    }

    public void setStopOnTerminate(Boolean stopOnTerminate) {
        this.stopOnTerminate = stopOnTerminate;
    }

    public ServiceProvider getServiceProvider() {
        return this.serviceProvider;
    }

    public void setServiceProvider(Integer providerId) {
        this.serviceProvider = ServiceProvider.forInt(providerId);
    }

    public void setServiceProvider(ServiceProvider provider) {
        this.serviceProvider = provider;
    }

    public Integer getInterval() {
        return interval;
    }

    public void setInterval(Integer interval) {
        this.interval = interval;
    }

    public Integer getFastestInterval() {
        return fastestInterval;
    }

    public void setFastestInterval(Integer fastestInterval) {
        this.fastestInterval = fastestInterval;
    }

    public Integer getActivitiesInterval() {
        return activitiesInterval;
    }

    public void setActivitiesInterval(Integer activitiesInterval) {
        this.activitiesInterval = activitiesInterval;
    }

    public String getLargeNotificationIcon () {
        String iconName = getNotificationIcon();
        if (iconName != null) {
            iconName = iconName + "_large";
        }
        return iconName;
    }

    public String getSmallNotificationIcon () {
        String iconName = getNotificationIcon();
        if (iconName != null) {
            iconName = iconName + "_small";
        }
        return iconName;
    }

    private void setActivityType(String activityType) {
        this.activityType = activityType;
    }

    private String getActivityType() {
        return activityType;
    }

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public String getHeaders() {
        return headers;
    }

    public void setHeaders(String headers) {
        this.headers = headers;
    }

    public String getParams() {
        return params;
    }

    public void setParams(String params) {
        this.params = params;
    }

    public void setOauthUrl(String oauthUrl) {
        this.oauthUrl = oauthUrl;
    }

    public String getOauthUrl() {
        return oauthUrl;
    }

    public String getAccessToken() {
        return accessToken;
    }

    public void setAccessToken(String accessToken) {
        this.accessToken = accessToken;
    }

    public String getRefreshToken() {
        return refreshToken;
    }

    public void setRefreshToken(String refreshToken) {
        this.refreshToken = refreshToken;
    }

    public String getClientId() {
        return clientId;
    }

    public void setClientId(String clientId) {
        this.clientId = clientId;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    @Override
    public String toString() {
        Map<String, Object> config = new HashMap<String, Object>();
        config.put("stationaryRadius", getStationaryRadius());
        config.put("desiredAccuracy", getDesiredAccuracy());
        config.put("distanceFilter", getDistanceFilter());
        config.put("locationTimeout", getLocationTimeout());
        config.put("debugging", isDebugging());
        config.put("notificationIcon", getNotificationIcon());
        config.put("notificationIconColor", getNotificationIconColor());
        config.put("notificationTitle", getNotificationTitle());
        config.put("notificationText", getNotificationText());
        config.put("stopOnTerminate", getStopOnTerminate());
        config.put("serviceProvider", getServiceProvider());
        config.put("interval", getInterval());
        config.put("fastestInterval", getFastestInterval());
        config.put("activitiesInterval", getActivitiesInterval());
        config.put("url", getUrl());
        config.put("headers", getHeaders());
        config.put("params", getParams());
        config.put("oauthUrl", getOauthUrl());
        config.put("accessToken", getAccessToken());
        config.put("refreshToken", getRefreshToken());
        config.put("clientId", getClientId());
        config.put("username", getUsername());
        config.put("password", getPassword());

        JSONObject json = new JSONObject(config);
        return json.toString();
    }

    public Parcel toParcel () {
        Parcel parcel = Parcel.obtain();
        this.writeToParcel(parcel, 0);
        parcel.setDataPosition(0);
        return parcel;
    }

    public static Config fromByteArray (byte[] byteArray) {
        Parcel parcel = Parcel.obtain();
        parcel.unmarshall(byteArray, 0, byteArray.length);
        parcel.setDataPosition(0);
        return Config.CREATOR.createFromParcel(parcel);
    }

    public static Config fromJSONArray (JSONArray data) throws JSONException {
        Config config = new Config();
        if (!data.isNull(0))
            config.setStationaryRadius((float) data.getDouble(0));
        if (!data.isNull(1))
            config.setDistanceFilter(data.getInt(1));
        if (!data.isNull(2))
            config.setLocationTimeout(data.getInt(2));
        if (!data.isNull(3))
            config.setDesiredAccuracy(data.getInt(3));
        if (!data.isNull(4))
            config.setDebugging(data.getBoolean(4));
        if (!data.isNull(5))
            config.setNotificationTitle(data.getString(5));
        if (!data.isNull(6))
            config.setNotificationText(data.getString(6));
        if (!data.isNull(7))
            config.setActivityType(data.getString(7));
        if (!data.isNull(8))
            config.setStopOnTerminate(data.getBoolean(8));
        if (!data.isNull(9))
            config.setNotificationIcon(data.getString(9));
        if (!data.isNull(10))
            config.setNotificationIconColor(data.getString(10));
        if (!data.isNull(11))
            config.setServiceProvider(data.getInt(11));
        if (!data.isNull(12))
            config.setInterval(data.getInt(12));
        if (!data.isNull(13))
            config.setFastestInterval(data.getInt(13));
        if (!data.isNull(14))
            config.setActivitiesInterval(data.getInt(14));
        if (!data.isNull(15))
            config.setUrl(data.getString(15));
        if (!data.isNull(16))
            config.setHeaders(data.getString(16));
        if (!data.isNull(17))
            config.setParams(data.getString(17));
        if (!data.isNull(18))
            config.setOauthUrl(data.getString(18));
        if (!data.isNull(19))
            config.setAccessToken(data.getString(19));
        if (!data.isNull(20))
            config.setRefreshToken(data.getString(20));
        if (!data.isNull(21))
            config.setClientId(data.getString(21));
        if (!data.isNull(22))
            config.setUsername(data.getString(22));
        if (!data.isNull(23))
            config.setPassword(data.getString(23));

        return config;
    }
}
